### PR TITLE
feat(arch): add support for wasm32 with simd128 feature

### LINKED ISSUE #217: WASM support
Whilst WASM doesn't expose the native CPU instructions, it does have its own SIMD intrinsics (see https://doc.rust-lang.org/core/arch/wasm32/index.html#simd).

I had a quick look at porting this crate to WASM this evening and it seems to be feasible. I've not had a chance to test it in the slightest, but it builds. (https://github.com/sam-kirby/simd-json/tree/wasm)

I'll try including it in a yew project I'm working on tomorrow and see if it a) works and b) offers any performance improvements.

_Originally posted by @sam-kirby in https://github.com/simd-lite/simd-json/issues/161#issuecomment-1043640140_

### PR DESCRIPTION
**Pull request by sam-kirby**
This adds support for two targets: 
1) `wasm32-unknown-unknown` - the current go to for running Rust on web browsers. The standard library is incredibly limited here and running tests is difficult.
2) `wasm32-wasi` - currently most useful for running tests, but may become more desirable in the future due to powerful sandboxing features offered by wasm runtimes on server.

In theory, any future wasm32 targets will be supported without any modification - e.g. `wasm32-web`. I've not made any effort to support `wasm64` yet, though enabling this in the future should be a simple matter of changing some cfg macros.

Some key changes and issues:
- proptest is incompatible with wasm targets as it relies on being able to fork, which is not possible. As a result, this dev dependency has to be gated and tests using it turned off on wasm targets. Due to `unused_imports` being a denied lint, certain imports in tests had to be broken up to allow compilation of tests on wasm32.
- testing requires wasmtime; a make rule has been added to make this as easy as possible. This could be substituted for any other wasm/wasi runtime.
- simdutf8 is currently a git dependency; support for wasm has not yet been included in a release of this crate
- value-trait also needed modification to support this, and I will open a separate PR there. Due to breaking changes this PR will target value-trait's 0.2.1 tag, not the current head.
- benchmarking doesn't work as libc ends up in the dependency graph somehow - I've not looked into this yet.
- whilst comparing the implementation for wasm to the neon version I noticed that the `vtstq_u8` intrinsic is now part of the standard library and so removed the temporary implementation. Do let me know if you'd like this reverted.
- wasm currently lacks any kind of runtime feature detection (though this should be possible in the future through [exception handling](https://github.com/WebAssembly/exception-handling/blob/master/proposals/exception-handling/Exceptions.md)), and simd instructions may be disabled (though in most cases they seem to be enabled by default). This may cause issues when considering the `allow-non-simd` feature.

closes #217

### CODE REVIEW

**Comment by Licenser** [comment]
Heya that is some great stuff!

I merged the value-trait PR it's tagged as 0.2.10 so that temporary dependency can go away, once the other one PR is merged I think we can merge this one too :) 

I'm quite excited, WASM support is really great! Thank you!

**Comment by Licenser** [comment]
I suspect the remaining overwrite is because we're waiting for a new release of simdutf8?

**Comment by sam-kirby** [comment]
That is correct, yes

**Comment by hkratz** [comment]
I will do a new release of simdutf8 shortly.

**Comment by hkratz** [comment]
simdutf8 v0.1.4 with WASM support has been released.

**Review by Licenser — APPROVED** [review]
Awesome, thank you!

**Comment by Licenser** [comment]
Weeeeh!

**Pull request merged**
